File: SetA_Question3.py
#!/usr/bin/env python3

'''
    This program selects a subset of patients from a data file based on
    their diagnosis. It gives the number of patients with diagnoses in
    different groups in the same month. For Cobbs Creek Healthcare Data
    Analysis Assessment Set A - Question #3
'''

import csv      # used to create reader objects from csv files
import sys      # used to accept command line arguments


def getGroup(claim_code):
    """
    This function converts a claim code into a group id
    :param claim_code: claim code of the patient
    :return: group that the patient fits in, or false if not in cohort
    """
    if claim_code in ["C83.0", "C83.00", "C83.01", "C83.02", "C83.03"]:
        return "A"
    elif claim_code in ["C91", "C91.1", "C91.10", "C91.11", "C91.12"]:
        return "B"
    elif claim_code in ["C95.10", "C95.90"]:
        return "C"
    else:
        return False


def main(argv):
    # accept optional command line argument for a data file or
    # use default file: sample_dx.csv
    if len(argv) == 2:
        dataFile = argv[1]
        print("Analyzing", dataFile, "...")
    else:
        dataFile = "sample_dx.csv"
        print("Analyzing default file: sample_dx.csv ...")

    # open the data file and store file object in data_csv
    data_csv = open(dataFile, "r")

    # create a reader object for the data file
    data = csv.reader(data_csv)

    cohort = {}             # create a dictionary for the cohort
    multiDiagnoses = set()  # create a set for patients with multiple diagnoses

    for r in data:
        # set group equal to the patient's diagnosis group and return false if no group
        if group := getGroup(r[2]):
            # check if patient id is alreacy in the cohort
            if r[0] in cohort:
                for d in cohort[r[0]]:
                    # if this patient has any different diagnoses in the same month
                    if r[3].partition("/")[0] in d and group not in d[r[3].partition("/")[0]]:
                        # add them to the multiple diagnoses set (sets ignore duplicates)
                        multiDiagnoses.add(r[0])
                        d[r[3].partition("/")[0]].append(group)
                        break
                # if there are no existing diagnoses in this monty, update the row
                if not any(r[3].partition("/")[0] in d for d in cohort[r[0]]):
                    cohort[r[0]].append({r[3].partition("/")[0]: [group]})
            # otherwise add the patient to the cohort
            else:
                cohort[r[0]] = [{r[3].partition("/")[0]: [group]}]
    # print cohort
    header = ["Cohort Patient IDs"]
    print(header[0])
    for r in cohort:
        print(r)

    # create CSV file for cohort
    with open("cohort.csv", 'w', newline='') as csvfile:
        csvwriter = csv.writer(csvfile)
        csvwriter.writerow(header)
        for r in cohort.keys():
            csvwriter.writerows([[r]])

    print("\nThe number of patients with diagnoses in different groups "
          "in the same month is:", len(multiDiagnoses),"\n")

    print("cohort.csv created\n")

    # Prevent window from closing
    input("press ENTER to exit")


if __name__ == '__main__':
    main(sys.argv)

File: SetA_Question3.R
#Author: Timothy Lentz
#version: R version 4.2.2 (2022-10-31 ucrt)
#Input: sample_dx.csv
#Output: cohort.csv

#open the data file
data = read.csv("sample_dx.csv")

#message the user
print("Analyzingsample_dx.csv...")
print("Producing cohort.csv...")

#Define the groups of diagnoses for the cohort
groupA = list("C83.0", "C83.00", "C83.01", "C83.02", "C83.03")
groupB = list("C91", "C91.1", "C91.10", "C91.11", "C91.12")
groupC = list("C95.10", "C95.90")

#Remove unnecessary data from the table
data["prescriber_id"] = NULL

#create subset of data with diagnoses from the cohort
data = subset(data, claim_code %in% c(groupA, groupB, groupC))

#remove duplicate patients from the cohort
cohort = data["patient_id"][!duplicated(data["patient_id"]), ]

#output a csv file with the names of patients in the cohort
patient_id = data.frame(cohort)
colnames(patient_id) = "Patient Id"
write.csv(patient_id, "cohort.csv", row.names=FALSE)

#replace diagnoses codes with their respective group
for(x in 1:length(groupA)){
  data["claim_code"][data["claim_code"] == groupA[x]] = "A"
}
for(x in 1:length(groupB)){
  data["claim_code"][data["claim_code"] == groupB[x]] = "B"
}
for(x in 1:length(groupC)){
  data["claim_code"][data["claim_code"] == groupC[x]] = "C"
}

#replace dates with their respective month
for (x in 1:nrow(data)){
  data[x, "service_date"] = substr(data[x, "service_date"],1,1)
}

#create a subset of the data for each group
dataA = subset(data, claim_code == "A")
dataB = subset(data, claim_code  == "B")
dataC = subset(data, claim_code == "C")

#merge each subset where patient_id and service_date is equal in both sets
AB = merge(dataA,dataB, by = c("patient_id","service_date"))
AC = merge(dataA,dataC, by = c("patient_id","service_date"))
BC = merge(dataB,dataC, by = c("patient_id","service_date"))

#combine the resulting sets, then remove duplicate patient_ids
total = rbind(AB["patient_id"],AC["patient_id"], BC["patient_id"])
total = total["patient_id"][!duplicated(total["patient_id"]), ]


print(c("The number of patients with diagnoses in different groups in the same month is: ", length(total)))
write.csv(patient_id, "cohort.csv", row.names=FALSE)

print("Done.")

Sys.sleep(15)
